README
NAME
    TOBYINK::Pod::HTML - convert Pod to HTML like TOBYINK

SYNOPSIS
       #!/usr/bin/perl
   
       use strict;
       use warnings;
       use TOBYINK::Pod::HTML;
   
       my $pod2html = "TOBYINK::Pod::HTML"->new(
          pretty             => 1,       # nicely indented HTML
          code_highlighting  => 1,       # use PPI::HTML
          code_line_numbers  => undef,
          code_styles        => {        # some CSS
             comment   => 'color:green',
             keyword   => 'font-weight:bold',
          }
       );
   
       print $pod2html->file_to_html(__FILE__);

DESCRIPTION
    Yet another pod2html converter.

    Note that this module requires Perl 5.14, and I have no interest in
    supporting legacy versions of Perl.

  Constructor
    `new(%attrs)`
        Moose-style constructor.

  Attributes
    `pretty`
        If true, will output pretty-printed (nicely indented) HTML. This
        doesn't make any difference to the appearance of the HTML in a
        browser.

        This feature requires XML::LibXML::PrettyPrint.

        Defaults to false.

    `code_highlighting`
        If true, source code samples within pod will be syntax highlighted as
        Perl 5.

        This feature requires PPI::HTML and PPI::Document.

        Defaults to false.

    `code_line_numbers`
        If undef, source code samples within pod will have line numbers, but
        only if they begin with "#!".

        If true, all source code samples within pod will have line numbers.

        This feature only works in conjunction with `code_highlighting`.

        Defaults to undef.

    `code_styles`
        A hashref of CSS styles to assign to highlighted code. The defaults
        are:

           +{
              symbol        => 'color:#333;background-color:#fcc',
              pod           => 'color:#060',
              comment       => 'color:#060;font-style:italic',
              operator      => 'color:#000;font-weight:bold',
              single        => 'color:#909',
              double        => 'color:#909',
              literal       => 'color:#909',
              interpolate   => 'color:#909',
              words         => 'color:#333;background-color:#ffc',
              regex         => 'color:#333;background-color:#9f9',
              match         => 'color:#333;background-color:#9f9',
              substitute    => 'color:#333;background-color:#f90',
              transliterate => 'color:#333;background-color:#f90',
              number        => 'color:#39C',
              magic         => 'color:#900;font-weight:bold',
              cast          => 'color:#f00;font-weight:bold',
              pragma        => 'color:#009',
              keyword       => 'color:#009;font-weight:bold',
              core          => 'color:#009;font-weight:bold',
              line_number   => 'color:#666',
              # for non-Perl code
              alert         => 'color:#f00;background-color:#ff0',
              warning       => 'color:#f00;background-color:#ff0;font-style:italic',
              error         => 'color:#f00;background-color:#ff0;font-style:italic;font-weight:bold',
              bstring       => '',
              function      => '',
              regionmarker  => '',
              others        => '',
           }

        Which looks kind of like the Perl highlighting from SciTE.

  Methods
    `file_to_dom($filename)`
        Convert pod from file to a XML::LibXML::Document object.

    `string_to_dom($document)`
        Convert pod from string to a XML::LibXML::Document object.

    `file_to_xhtml($filename)`
        Convert pod from file to an XHTML string.

    `string_to_xhtml($document)`
        Convert pod from string to an XHTML string.

    `file_to_html($filename)`
        Convert pod from file to an HTML5 string.

        This feature requires HTML::HTML5::Writer.

    `string_to_html($document)`
        Convert pod from string to an HTML5 string.

        This feature requires HTML::HTML5::Writer.

  Alternative Syntax Highlighting
    This module defines an additional Pod command to change the language for
    syntax highlighting. To tell TOBYINK::Pod::HTML to switch to, say, Haskell
    instead of the default (Perl), just use:

       =for highlighter language=Haskell

    Then all subsequent code samples will be highlighted as Haskell, until
    another such command is seen.

    While syntax highlighting for Perl uses PPI::HTML, syntax highlighting for
    other languages uses either Syntax::Highlight::RDF or
    Syntax::Highlight::Engine::Kate as appropriate, so you need to have them
    installed if you want this feature.

    Note that the language names defined by Syntax::Highlight::Engine::Kate
    are case-sensitive, and TOBYINK::Pod::HTML makes no attempt at
    case-folding, so you must use the correct case!

    Note that only the PPI highlighter supports line numbering.

    The following command can be used to switch to plain text syntax
    highlighting (i.e. no highlighting at all):

       =for highlighter language=Text

SEE ALSO
    Pod::Simple, PPI::HTML, etc.

AUTHOR
    Toby Inkster <tobyink@cpan.org>.

COPYRIGHT AND LICENCE
    This software is copyright (c) 2013 by Toby Inkster.

    This is free software; you can redistribute it and/or modify it under the
    same terms as the Perl 5 programming language system itself.

DISCLAIMER OF WARRANTIES
    THIS PACKAGE IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
    WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
    MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.



Changes
TOBYINK-Pod-HTML
================

Created:      2013-01-18
Home page:    <https://metacpan.org/release/TOBYINK-Pod-HTML>
Bug tracker:  <http://rt.cpan.org/Dist/Display.html?Queue=TOBYINK-Pod-HTML>
Maintainer:   Toby Inkster (TOBYINK) <tobyink@cpan.org>

0.004	2013-11-18

 [ Bug Fixes ]
 - Don't need to use <br> within <pre>...</pre>.

 [ Packaging ]
 - use Dist::Inkt

0.003	2013-04-10

 [ Bug Fixes ]
 - Switching syntax highlight language in documents that have no NAME
   section was broken due to Pod::Simple::PullParser's look-ahead feature.
   This should now work.

 [ Other ]
 - Added: Syntax highlighting for Semantic Web related languages using
   Syntax::Highlight::RDF.

0.002	2013-03-21

 - Added: Allow non-Perl languages to be syntax highlighted too.

0.001	2013-02-26	Initial release


Makefile.PL
use strict;
use ExtUtils::MakeMaker 6.17;

my $EUMM = eval( $ExtUtils::MakeMaker::VERSION );

my $meta = {
  "abstract"           => "convert Pod to HTML like TOBYINK",
  "author"             => ["Toby Inkster (TOBYINK) <tobyink\@cpan.org>"],
  "dynamic_config"     => 0,
  "generated_by"       => "Dist::Inkt::Profile::TOBYINK version 0.013, CPAN::Meta::Converter version 2.120921",
  "keywords"           => [],
  "license"            => ["perl_5"],
  "meta-spec"          => {
                            url => "http://search.cpan.org/perldoc?CPAN::Meta::Spec",
                            version => 2,
                          },
  "name"               => "TOBYINK-Pod-HTML",
  "no_index"           => { directory => ["eg", "examples", "inc", "t", "xt"] },
  "prereqs"            => {
                            configure => { requires => { "ExtUtils::MakeMaker" => 6.17 } },
                            runtime   => {
                                           recommends => {
                                                           "HTML::HTML5::Writer"             => 0.201,
                                                           "PPI::Document"                   => 0,
                                                           "PPI::HTML"                       => 0,
                                                           "Syntax::Highlight::Engine::Kate" => 0,
                                                           "Syntax::Highlight::RDF"          => 0,
                                                           "XML::LibXML::PrettyPrint"        => 0,
                                                         },
                                           requires   => {
                                                           "HTML::HTML5::Parser" => 0.208,
                                                           "Moo" => 1.000008,
                                                           "XML::LibXML::QuerySelector" => 0.003,
                                                         },
                                         },
                          },
  "provides"           => {
                            "TOBYINK::Pod::HTML"         => { file => "lib/TOBYINK/Pod/HTML.pm", version => 0.004 },
                            "TOBYINK::Pod::HTML::Helper" => { file => "lib/TOBYINK/Pod/HTML.pm", version => 0.004 },
                          },
  "release_status"     => "stable",
  "resources"          => {
                            bugtracker   => {
                                              web => "http://rt.cpan.org/Dist/Display.html?Queue=TOBYINK-Pod-HTML",
                                            },
                            homepage     => "https://metacpan.org/release/TOBYINK-Pod-HTML",
                            license      => ["http://dev.perl.org/licenses/"],
                            repository   => {
                                              type => "git",
                                              web  => "https://github.com/tobyink/p5-tobyink-pod-html",
                                            },
                            X_identifier => "http://purl.org/NET/cpan-uri/dist/TOBYINK-Pod-HTML/project",
                          },
  "version"            => 0.004,
  "x_provides_scripts" => { "pod2html-tobyink" => { file => "bin/pod2html-tobyink" } },
};

my %dynamic_config;

my %WriteMakefileArgs = (
	ABSTRACT   => $meta->{abstract},
	AUTHOR     => ($EUMM >= 6.5702 ? $meta->{author} : $meta->{author}[0]),
	DISTNAME   => $meta->{name},
	VERSION    => $meta->{version},
	EXE_FILES  => [ map $_->{file}, values %{ $meta->{x_provides_scripts} || {} } ],
	NAME       => do { my $n = $meta->{name}; $n =~ s/-/::/g; $n },
	%dynamic_config,
);

$WriteMakefileArgs{LICENSE} = $meta->{license}[0] if $EUMM >= 6.3001;

sub deps
{
	my %r;
	for my $stage (@_)
	{
		for my $dep (keys %{$meta->{prereqs}{$stage}{requires}})
		{
			next if $dep eq 'perl';
			my $ver = $meta->{prereqs}{$stage}{requires}{$dep};
			$r{$dep} = $ver if !exists($r{$dep}) || $ver >= $r{$dep};
		}
	}
	\%r;
}

my ($build_requires, $configure_requires, $runtime_requires, $test_requires);
if ($EUMM >= 6.6303)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{TEST_REQUIRES}      ||= deps('test');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');
}
elsif ($EUMM >= 6.5503)
{
	$WriteMakefileArgs{BUILD_REQUIRES}     ||= deps('build', 'test');
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime');	
}
elsif ($EUMM >= 6.52)
{
	$WriteMakefileArgs{CONFIGURE_REQUIRES} ||= deps('configure');
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('runtime', 'build', 'test');	
}
else
{
	$WriteMakefileArgs{PREREQ_PM}          ||= deps('configure', 'build', 'test', 'runtime');	
}

{
	my ($minperl) = reverse sort(
		grep defined && /^[0-9]+(\.[0-9]+)?$/,
		map $meta->{prereqs}{$_}{requires}{perl},
		qw( configure build runtime )
	);
	
	if (defined($minperl))
	{
		die "Installing $meta->{name} requires Perl >= $minperl"
			unless $] >= $minperl;
		
		$WriteMakefileArgs{MIN_PERL_VERSION} ||= $minperl
			if $EUMM >= 6.48;
	}
}

sub FixMakefile
{
	return unless -d 'inc';
	my $file = shift;
	
	local *MAKEFILE;
	open MAKEFILE, "< $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	my $makefile = do { local $/; <MAKEFILE> };
	close MAKEFILE or die $!;
	
	$makefile =~ s/\b(test_harness\(\$\(TEST_VERBOSE\), )/$1'inc', /;
	$makefile =~ s/( -I\$\(INST_ARCHLIB\))/ -Iinc$1/g;
	$makefile =~ s/( "-I\$\(INST_LIB\)")/ "-Iinc"$1/g;
	$makefile =~ s/^(FULLPERL = .*)/$1 "-Iinc"/m;
	$makefile =~ s/^(PERL = .*)/$1 "-Iinc"/m;
	
	open  MAKEFILE, "> $file" or die "FixMakefile: Couldn't open $file: $!; bailing out";
	print MAKEFILE $makefile or die $!;
	close MAKEFILE or die $!;
}

my $mm = WriteMakefile(%WriteMakefileArgs);
FixMakefile($mm->{FIRST_MAKEFILE} || 'Makefile');
exit(0);

